R/file_helpers.R

```diff
@@ -1,122 +1,188 @@
-      dir(temp_dir, recursive = TRUE, full.names = TRUE)
-    )
+    temp_dir            <- file.path(tempdir(), "data")
+    remote_files        <- copy_data_files(remote_files, temp_dir)
+    remote_files        <- normalizePath(remote_files)
     names(remote_files) <- dir(temp_dir, recursive = TRUE)
   }
 
-  # Combine paths to local files and remote files
   files <- c(local_files, remote_files)
 
   # Specify full paths in chunk options
+  set_option_exercise_files(files)
+
+  invisible(files)
+}
+
+find_local_files <- function() {
+  data_contents <- dir("data", recursive = TRUE)
+
+  if (is.null(data_contents)) {return(NULL)}
+
+  # Generate paths to files in the `data/` directory
+  local_files        <- normalizePath(file.path("data", data_contents))
+  names(local_files) <- data_contents
+
+  local_files
+}
+
+set_option_exercise_files <- function(files) {
   knitr::opts_chunk$set(exercise.files = files)
 }
 
+get_option_exercise_files <- function(exercise = NULL) {
+  default_files <- knitr::opts_chunk$get("exercise.files")
+
+  manual_exercise_files <- exercise$options$exercise.files
+  manual_remote_files   <- exercise$options$exercise.remote.files
+
+  if (!is.null(c(manual_exercise_files, manual_remote_files))) {
+    warning(
+      paste(
+        "Manually setting the `exercise.files` or `exercise.remote.files`",
+        "chunk options causes suboptimal performance.\nPlease see",
+        "`?learnr::use_remote_files` for more information."
+      )
+    )
+  }
+
+  c(default_files, manual_exercise_files, manual_remote_files)
+}
+
+set_option_remote_files <- function(files) {
+  if (any(!is_url(files) & !is_system_file(files))) {
+    stop(
+      "Remote files must be either URLs or the result of a call to ",
+      "`system.file()`.",
+      call. = FALSE
+    )
+  }
+
+  knitr::opts_chunk$set(exercise.remote.files = files)
+}
+
+get_option_remote_files <- function() {
+  knitr::opts_chunk$get("exercise.remote.files")
+}
+
 copy_data_files <- function(files, dest_dir = "data") {
+  if (!length(files)) {return(NULL)}
+
   # If a name is not specified, default to the basename of the file
   names(files)[names(files) == ""] <- basename(files[names(files) == ""])
 
-  # Paths files will be copied into
   dest_path <- file.path(dest_dir, names(files))
 
   for (i in seq_along(files)) {
-    # Create directory structure
+    # Allow authors to add files in nested directories
     dir.create(dirname(dest_path[[i]]), showWarnings = FALSE, recursive = TRUE)
-
-    # Copy files
     copy_file(files[[i]], dest_path[[i]])
   }
 
-  files
+  invisible(dest_path)
 }
 
 copy_file <- function(from, to, ...) {
   if (is_url(from)) {
-    download.file(from, to, ...)
+    utils::download.file(from, to, ...)
   } else {
     file.copy(from, to, ...)
   }
 }
 
+is_system_file <- function(path) {
+  apply(
+    as.matrix(
+      vapply(.libPaths(), grepl, logical(length(path)), path, fixed = TRUE)
+    ),
+    1,
+    any
+  )
```

Comment:
This is fun! But also not very readable. I played with it a bit and decided I like this version slightly better:

```suggestion
  vapply(
    lapply(path, Vectorize(grepl, "pattern"), pattern = .libPaths(), fixed = TRUE),
    any,
    logical(1)
  )
```

but...

Reply:
...but I think it might be even better as a slightly recursive function

```r
is_system_file <- function(path) {
  if (length(path) > 1) {
    vapply(path, is_system_file, logical(1))
  }
  any(vapply(.libPaths(), grepl, x = path, fixed = TRUE, logical(1)))
}	
```

Reply:
Love this!